[RFC] Create functional aten assertion ops

XuanQi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):

* #103887
* #103757
* __->__ #103751

Prep PR to create functional version of assertions. Concrete logic will be implemented in future PRs.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78

XuanQi committed:
[RFC] Create functional aten assertion ops

[ghstack-poisoned]

XuanQi committed:
Update on "[RFC] Create functional aten assertion ops"




[ghstack-poisoned]

XuanQi committed:
Update on "[RFC] Create functional aten assertion ops"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

XuanQi committed:
Update on "[RFC] Create functional aten assertion ops"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

XuanQi committed:
Update on "[RFC] Create functional aten assertion ops"


Prep PR to create functional version of assertions. Concrete logic will be implemented in future PRs.

[ghstack-poisoned]

tugsbayasgalan commented on aten/src/ATen/native/Constraints.cpp:
```diff
@@ -11,5 +12,13 @@ void sym_constrain_range_cpu(
+    c10::optional<int64_t> min = c10::nullopt,
+    c10::optional<int64_t> max = c10::nullopt,
+    const c10::optional<Tensor>& dep_token = c10::nullopt) {
+  return c10::value_or_else(dep_token, []{return Tensor();});
```
I think we should create a empty tensor and return: at::empty({})?

tugsbayasgalan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -170,18 +170,30 @@
     CPU: sym_constrain_range_cpu
     CUDA: sym_constrain_range_cuda
 
+- func: functional_sym_constrain_range(Scalar size, int? min=None, int? max=None, Tensor? dep_token=None) -> Tensor
```
Why does it have take optional dep_token?

XuanQi replied:
Yeah this is a little bit annoying but since both `min` and `max` before it are both optional so I make it optional as well. The other option is to just move it before `min` / `max` arg but I feel like it makes it a little bit inconsistency with other functional assertions since `dep_token` is the last in others.
Not idea but let me add a runtime check for it to make sure `dep_token` is passed in?

tugsbayasgalan replied:
Haha I see what you mean. Hmm I think it is fine to have it before the min/max. I feel like we shouldn't make the dep_token implementation more complicated.

XuanQi committed:
Update on "[RFC] Create functional aten assertion ops"


Prep PR to create functional version of assertions. Concrete logic will be implemented in future PRs.

[ghstack-poisoned]

XuanQi committed:
Update on "[RFC] Create functional aten assertion ops"


Prep PR to create functional version of assertions. Concrete logic will be implemented in future PRs.

[ghstack-poisoned]

XuanQi committed:
Update on "[RFC] Create functional aten assertion ops"

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):

* #103887
* #103757
* __->__ #103751

Prep PR to create functional version of assertions. Concrete logic will be implemented in future PRs.

[ghstack-poisoned]

XuanQi committed:
Update on "[RFC] Create functional aten assertion ops"

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):

* #103887
* #103757
* __->__ #103751

Prep PR to create functional version of assertions. Concrete logic will be implemented in future PRs.

[ghstack-poisoned]

tugsbayasgalan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -170,18 +170,34 @@
+  dispatch:
+    CPU: functional_sym_constrain_range_cpu
+
+- func: make_dep_token(Tensor? dep_token) -> Tensor
```
Why does it need a optional dep_token?

XuanQi replied:
Got some trouble defining op without any argument. But I think got a version which is kind of working. Will update.

XuanQi committed:
Update on "[RFC] Create functional aten assertion ops"

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):

* #103887
* #103757
* __->__ #103751

Prep PR to create functional version of assertions. Concrete logic will be implemented in future PRs.

[ghstack-poisoned]

tugsbayasgalan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -170,18 +170,34 @@
     CPU: _assert_async_cpu
     CUDA: _assert_async_cuda
 
+- func: _functional_assert_async(Tensor self, Tensor dep_token) -> Tensor
```
I don't think we actually need this. Dynamo always uses the _assert_async.msg variant.

tugsbayasgalan commented on aten/src/ATen/native/Constraints.cpp:
```diff
@@ -1,15 +1,50 @@
+    c10::optional<bool> pin_memory_opt,
+    c10::optional<c10::MemoryFormat> memory_format_opt) {
+  return at::empty(
+      {0},
```
I don't think we need to pass 0 here. Just at::empty({}) should be good enough.

XuanQi committed:
Update on "[RFC] Create functional aten assertion ops"

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):

* #103887
* #103757
* __->__ #103751

Prep PR to create functional version of assertions. Concrete logic will be implemented in future PRs.

[ghstack-poisoned]

XuanQi committed:
Update on "[RFC] Create functional aten assertion ops"

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):

* #103887
* #103757
* __->__ #103751

Prep PR to create functional version of assertions. Concrete logic will be implemented in future PRs.

[ghstack-poisoned]

XuanQi committed:
Update on "[RFC] Create functional aten assertion ops"

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):

* #103887
* #103757
* __->__ #103751

Prep PR to create functional version of assertions. Concrete logic will be implemented in future PRs.

[ghstack-poisoned]

XuanQi commented:
@pytorchbot merge

XuanQi commented:
@pytorchbot label "topic: not user facing"

XuanQi commented:
@pytorchbot merge

XuanQi committed:
Update on "[RFC] Create functional aten assertion ops"

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):

* #103887
* #103757
* __->__ #103751

Prep PR to create functional version of assertions. Concrete logic will be implemented in future PRs.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

XuanQi commented:
@pytorchbot merge

zou3519 commented on aten/src/ATen/native/TensorCompare.cpp:
```diff
@@ -409,6 +410,15 @@ void _assert_async_msg_cpu(const Tensor& self, c10::string_view assert_msg) {
   TORCH_CHECK(native::is_nonzero(self), assert_msg != "" ? assert_msg : "Assertion is failed");
 }
 
+Tensor _functional_assert_async_msg_cpu(
+  const Tensor& self,
+  c10::string_view assert_msg,
+  const Tensor& dep_token) {
+  _assert_async_msg_cpu(self, assert_msg);
+  return dep_token;
+}
```
Mentioned this over chat to Tugsuu but just to point out the problem: PyTorch has a problem if your non-CompositeImplicitAutograd operator has incorrect alias information. The schema promised that the operator was functional (and that none of the outputs alias the inputs), but we are returning an input (the world token) directly as an output.

This can end up triggering asserts (some of which can only be see in a DEBUG build).

The fix for this is to just `return dep_token.clone()`. The dep_token should be a Tensor without storage anyways, so the clone should be cheap.
